### ISSUE TITLE
stubbing real object is silently ignored

### ISSUE THREAD

**Issue by Krever** [mockito/mockito-scala#102]
Following code throws a nice exception as expected:
```
import org.mockito.IdiomaticMockito
import org.scalatest.FunSuite

class Foo {
  def bar(x: String): String = x
}

class TestTest extends FunSuite with IdiomaticMockito {
  val x = new Foo
  x.bar("x").shouldReturn("XXXX")
}
```
But if we add at least one successfully created mock at the beginning it runs ok:
```
import org.mockito.IdiomaticMockito
import org.scalatest.FunSuite

class Foo {
  def bar(x: String): String = x
}

class TestTest extends FunSuite with IdiomaticMockito {
  val y = mock[Foo]
  y.toString // fool warn-unused

  val x = new Foo
  x.bar("x").shouldReturn("XXXX")
}
```
It was extremely hard to identify the problem and minimize :D

**Comment by ultrasecreth** [comment]
Interesting, does the same happen with the old syntax?

**Comment by ultrasecreth** [comment]
Ok, the issue there is that the line `y.toString` is the one being stubbed. If you add `println(y.toString)` at the end of the test it will print `XXXX`.

so if you remove `y.toString` you'll get the exception you're looking for ;)

**Comment by Krever** [comment]
I does indeed! Do you know why this happens? Very peculiar behavior IMHO.

**Comment by ultrasecreth** [comment]
The thing is that mockito mainly works by using (abusing (?)) the substitution model.
In this case this means that it relies on the fact that whatever you pass to the `when` method will be called before the `when` method itself, but mockito doesn't actually know what has been executed, instead, each mock records the calls it has received and when the `when` method is called it checks if some mock has a method call that may be a stubbing.
So in this case, the call to `y.toString` puts that mock in "they may be stubbing me" state, and then when you call `when` it thinks that that's the method you're trying to stub.
Basically you got a lot of stars aligned to get this working:

- there is only one invocation on a mock
- both methods return string
- the object called inside the `when` is not a mock

Makes sense?

**Comment by Krever** [comment]
It makes sense now! Do you see some way to protect against this "feature" of mockito in mockito-scala?
Maybe we could add a check somehwere before calling Mockito.

**Comment by ultrasecreth** [comment]
Sadly I don't think we have more tools in Scala than in Java to help us to prevent this kinda things.
I've been thinking for a while to see if I can make the API a bit more type safe, but so far I haven't found anything that works in all scenarios and doesn't add extra complexity.
But be sure if I can think of something I'll do it!
